blockchain/indexers/common.go

```diff
@@ -173,3 +182,900 @@ func interruptRequested(ctx context.Context) bool {
+	// subscription. It is expected that the subscriber associated with the
+	// prerequisite provided processes notifications before they are
+	// delivered by this subscription to its subscriber. An empty string
+	// indicates the subscription has no prerequisite.
+	prerequisite string
+
+	// dependent defines the index subscription that requires the subscriber
+	// associated with this subscription to have processed incoming
+	// notifications before it does. A nil dependency indicates the subscription
+	// has no dependencies.
+	dependent *IndexSubscription
+}
+
+// newIndexSubscription initializes a new index subscription.
+func newIndexSubscription(subber *IndexSubscriber, indexer Indexer, prereq string) *IndexSubscription {
+	return &IndexSubscription{
+		id:           indexer.Name(),
+		idx:          indexer,
+		prerequisite: prereq,
+		subscriber:   subber,
+		c:            make(chan *IndexNtfn, bufferSize),
+	}
+}
+
+// C returns a channel that produces a stream of index notifications.
+// Successive calls to C return the same channel.
+func (s *IndexSubscription) C() chan *IndexNtfn {
+	return s.c
+}
+
+// stop prevents any future index updates from being delivered and
+// unsubscribes the associated subscription.
+func (s *IndexSubscription) stop() error {
+
+	// If the subscription has a prerequisite, find it and remove the
+	// subscription as a dependency.
+	if s.prerequisite != noPrereqs {
+		s.mtx.Lock()
+		prereq, ok := s.subscriber.subscriptions[s.prerequisite]
+		s.mtx.Unlock()
+		if !ok {
+			return fmt.Errorf("no subscription found with id %s", s.prerequisite)
+		}
+
+		prereq.mtx.Lock()
+		prereq.dependent = nil
+		prereq.mtx.Unlock()
+
+		return nil
+	}
+
+	// If the subscription has a dependent, stop it as well.
+	if s.dependent != nil {
+		err := s.dependent.stop()
+		if err != nil {
+			return err
+		}
+	}
+
+	// If the subscription is independent, remove it from the
+	// index subscriber's subscriptions.
+	s.mtx.Lock()
+	delete(s.subscriber.subscriptions, s.id)
+	s.mtx.Unlock()
+
+	return nil
+}
+
+// updateIndex processes the notification for the provided index.
+func updateIndex(ctx context.Context, indexer Indexer, ntfn *IndexNtfn) error {
+	// Ensure the incoming index notification is of the next block
+	// extending the chain.
+	tip, _, err := indexer.Tip()
+	if err != nil {
+		return fmt.Errorf("%s: unable to fetch index tip: %v",
+			indexer.Name(), err)
+	}
+
+	var expectedHeight int64
+	switch ntfn.NtfnType {
+	case ConnectNtfn:
+		expectedHeight = tip + 1
+	case DisconnectNtfn:
+		expectedHeight = tip
+	default:
+		return fmt.Errorf("%s: unknown notification type received: %v",
+			indexer.Name(), ntfn.NtfnType)
+	}
+
+	// Relay the notification to the dependent if it is not expected
+	// notification.
+	if expectedHeight != ntfn.Block.Height() {
+		log.Tracef("%s: relaying notification for height %d to dependent",
+			indexer.Name(), ntfn.Block.Height())
+		return notifyDependent(ctx, indexer, ntfn)
```

Comment:
If the notification isn't for the expected height, why does it just notify the dependent and move on?  Wouldn't this result in a missed notification for this index?

Reply:
For index notifications that are less than the index tip height its fine to just relay the notification to the dependent since the dependent could be synced to a lower tip height. I do agree the check needs refinement, it needs to error if the index notification received is greater than the expected height.